src/v2/auth.rs

```diff
@@ -1,23 +1,165 @@
+#[derive(Debug, Clone)]
+pub struct BasicAuth {
+    user: String,
+    password: Option<String>,
+}
+
+/// Structured representation for the content of the authentication response header.
+#[derive(Debug, PartialEq, Eq, Deserialize)]
+pub(crate) enum WwwAuthenticateHeaderContent {
+    Bearer(WwwAuthenticateHeaderContentBearer),
+    Basic(WwwAuthenticateHeaderContentBasic),
+}
+
+impl WwwAuthenticateHeaderContent {
+    /// Create a `WwwAuthenticateHeaderContent` by parsing a `HeaderValue` instance.
+    pub(crate) fn from_www_authentication_header(header_value: HeaderValue) -> Result<Self> {
+        let header = String::from_utf8(header_value.as_bytes().to_vec())?;
+
+        // This regex will result in multiple captures which will contain one key-value pair each.
+        // The first capture will be the only one with the "method" group set.
+        let re = regex::Regex::new(
+            r#"(?x)\s*
+            ((?P<method>[A-Z][a-z]+)\s*)?
+            (
+                \s*
+                    (?P<key>[a-z]+)
+                \s*
+                    =
+                \s*
+                    "(?P<value>[^"]+)"
+                \s*
+            )
+        "#,
+        )?;
+        let captures = re.captures_iter(&header).collect::<Vec<_>>();
+
+        let method = captures
+            .get(0)
+            .ok_or_else(|| {
+                Error::from(format!("regex '{}' didn't match '{}'", re.as_str(), header))
+            })?
+            .name("method")
+            .ok_or_else(|| Error::from(format!("method not found in {}", header)))?
+            .as_str();
+
+        let serialized_content = serde_json::Value::Object(serde_json::Map::from_iter(
+            [(
+                method.to_string(),
+                serde_json::Value::Object(serde_json::Map::from_iter(
+                    captures.into_iter().filter_map(|capture| {
+                        match (
+                            capture.name("key").map(|n| n.as_str().to_string()),
+                            capture.name("value").map(|n| n.as_str().to_string()),
+                        ) {
+                            (Some(key), Some(value)) => {
+                                Some((key, serde_json::Value::String(value)))
+                            }
+                            _ => None,
+                        }
+                    }),
+                )),
+            )]
+            .iter()
+            .cloned(),
+        ))
+        .to_string();
+
+        // Deserialize the content
+        let mut unsupported_keys = std::collections::HashSet::new();
+        let content: WwwAuthenticateHeaderContent = serde_ignored::deserialize(
+            &mut serde_json::Deserializer::from_str(&serialized_content),
+            |path| {
+                unsupported_keys.insert(path.to_string());
+            },
+        )?;
+
+        if !unsupported_keys.is_empty() {
+            warn!("unsupported keys remaining {:#?}", unsupported_keys);
```

Comment:
Previously we threw an error in this case. I decided to change this because I think that as long as the parsing was successful, we may just work with the result and not be overly strict.

Reply:
Added https://github.com/camallo/dkregistry-rs/pull/159/commits/68da0dc2f657c2566ebbba8f927268b4c11e50e3 on top which implements the plain string route. I like the aspect that we don't need the intermediate `Value` and I can't spot any weakness in the approach. I'd appreciate thoughts here.

Reply:
I don't have strong opinions in either directions right now. If we stick to a warning log, we should add a bit of context like "skipping unrecognized keys in authentication header: {:#?}".